community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java

```diff
@@ -0,0 +1,584 @@
+import org.neo4j.kernel.KernelEventHandlers;
+import org.neo4j.kernel.PlaceboTransaction;
+import org.neo4j.kernel.TopLevelTransaction;
+import org.neo4j.kernel.TransactionEventHandlers;
+import org.neo4j.kernel.api.EntityType;
+import org.neo4j.kernel.api.KernelAPI;
+import org.neo4j.kernel.api.KernelTransaction;
+import org.neo4j.kernel.api.ReadOperations;
+import org.neo4j.kernel.api.Statement;
+import org.neo4j.kernel.api.exceptions.EntityNotFoundException;
+import org.neo4j.kernel.api.exceptions.InvalidTransactionTypeKernelException;
+import org.neo4j.kernel.api.exceptions.index.IndexNotFoundKernelException;
+import org.neo4j.kernel.api.exceptions.schema.ConstraintValidationKernelException;
+import org.neo4j.kernel.api.exceptions.schema.SchemaKernelException;
+import org.neo4j.kernel.api.exceptions.schema.SchemaRuleNotFoundException;
+import org.neo4j.kernel.api.index.IndexDescriptor;
+import org.neo4j.kernel.api.index.InternalIndexState;
+import org.neo4j.kernel.impl.api.operations.KeyReadOperations;
+import org.neo4j.kernel.impl.core.NodeManager;
+import org.neo4j.kernel.impl.core.ThreadToStatementContextBridge;
+import org.neo4j.kernel.impl.query.QueryEngineProvider;
+import org.neo4j.kernel.impl.query.QueryExecutionEngine;
+import org.neo4j.kernel.impl.query.QueryExecutionKernelException;
+import org.neo4j.kernel.impl.store.StoreId;
+import org.neo4j.kernel.impl.traversal.BidirectionalTraversalDescriptionImpl;
+import org.neo4j.kernel.impl.traversal.MonoDirectionalTraversalDescription;
+import org.neo4j.kernel.lifecycle.LifeSupport;
+import org.neo4j.kernel.lifecycle.LifecycleException;
+import org.neo4j.logging.Log;
+import org.neo4j.tooling.GlobalGraphOperations;
+
+import static java.lang.String.format;
+
+import static org.neo4j.collection.primitive.PrimitiveLongCollections.map;
+import static org.neo4j.helpers.collection.IteratorUtil.emptyIterator;
+import static org.neo4j.kernel.impl.api.operations.KeyReadOperations.NO_SUCH_LABEL;
+import static org.neo4j.kernel.impl.api.operations.KeyReadOperations.NO_SUCH_PROPERTY_KEY;
+
+/**
+ * Implementation of the GraphDatabaseService/GraphDatabaseService interfaces. This delegates to the
+ * services created by the {@link org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory}.
+ *
+ * To make a custom GraphDatabaseFacade, the best option is to subclass an existing GraphDatabaseFacadeFactory. Another
+ * alternative, used by legacy database implementations, is to subclass this class and call
+ * {@link org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory#newFacade(java.util.Map, org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.Dependencies, GraphDatabaseFacade)} in the constructor.
+ */
+public class GraphDatabaseFacade
+    implements GraphDatabaseAPI
+{
+    private static final long MAX_NODE_ID = IdType.NODE.getMaxValue();
+    private static final long MAX_RELATIONSHIP_ID = IdType.RELATIONSHIP.getMaxValue();
+
+    private boolean initialized = false;
```

Comment:
I can't quite see this flag serving a purpose. It is only assigned in the constructor, so it could be final. But the only cases when that assignment wouldn't take place are cases when the object wouldn't be published (because of exceptions in the constructor) so it does in fact serve no purpose.

Perhaps it should be traded for a flag that helps making shutdown idempotent?